File: algo/257.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def binaryTreePaths(self, root):
        """
        :type root: TreeNode
        :rtype: List[str]
        """
        if root is None:
            return []
        else:
            paths = []
            path_stack = []
            self.helper(root, paths, path_stack)
            return path_stack

    def helper(self, root, paths, path_stack):
        if root is None:
            return
        elif root.left is None and root.right is None:
            path_stack.append(str(root.val))
            paths.append("->".join(path_stack))
            path_stack.pop()
        else:
            path_stack.append(str(root.val))
            self.helper(root.left, paths, path_stack)
            self.helper(root.right, paths, path_stack)
            path_stack.pop()
        return

File: algo/257.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
#include "listnode.h"
#include <vector>
#include <algorithm>
#include <string>
#include <iostream>
#include <fstream>
#include <sstream>
#include <unordered_map>
#include <unordered_set>

using namespace std;
class Solution {
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        
    }

    bool isleaf(TreeNode* root) {
        return (root -> left == nullptr && root -> right == nullptr);
    }

    void add_path() {

    }

    void helper(TreeNode* root, vector<string>& paths, vector<int> path_stack) {
        if (isleaf(root)) {
            path_stack.push_back(root -> val);
            
            path_stack.pop_back();
        } else {
            path_stack.push_back(root -> val);
            helper(root -> left);
            helper(root -> right);
            path_stack.pop_back();
        }
    }
};